File: lab5/prolog/primes.pl
hasFactor(N, Factor) :-
    N mod Factor =:= 0.
hasFactor(N, Factor) :-
    Factor * Factor < N,
    NextFactor is Factor + 2
    hasFactor(N, NextFactor).

isPrime(2).
isPrime(3).
isPrime(P) :-
    P > 3,
    P mod 2 =\= 0,
    \+ hasFactor(P, 3).

primes(N, Primes) :-
    findall(P, (between(2, N, P), isPrime(P)), Primes).

File: lab5/haskel/primes.hs
primes :: Int -> [Int]
primes n
    | n < 2     = []
    | otherwise = sieve [2..n]
    where
        sieve :: [Int] -> [Int]
        sieve [] = []
        sieve (p:xs)
            | p * p > n = p : xs
            | otherwise = p : sieve [x | x <- xs, x `mod` p /= 0]